### PR TITLE
fix line end determination for a, A, $

### PR DESCRIPTION
**Pull request by kimitake**
fix line end determination for a, A, $

### CODE REVIEW

**Comment by jpoon** [comment]
I think it might be more proper to separate this into two classes. A cursor class (in which insert mode will use) and a caret class (in which command mode will use). 

As the same bug will appear for `lineEnd` and `documentEnd`

**Comment by kimitake** [comment]
just created Caret class which is derived from Cursor class, and defined "maxLineLength" method for override between Caret and Cursor class.
And normal mode uses Caret class instead of Cursor.
How about this?

File: test/cursor.test.ts [hunk 1]
```diff
@@ -56,13 +57,13 @@ suite("cursor", () => {
...
+		Caret.move(new vscode.Position(0, 7));
 
-		let current = Cursor.currentPosition();
+		let current = Caret.currentPosition();
 		assert.equal(current.line, 0);
 		assert.equal(current.character, 7);
 
-		let right = Cursor.right();
+		let right = Caret.right();
 		assert.equal(right.line, 0);
```
Comment (jpoon) [inline test/cursor.test.ts]: Can we keep this test suite focused on the `Cursor` class? Keep using the Cursor functions but change the expected values on the asserts?

**Comment by jpoon** [comment]
:+1: Just one note. After that, we can mergeeeeee!

**Comment by kimitake** [comment]
Is this correct?

**Pull request merged**